src/ngx_http_lua_kong_ssl.c

```diff
@@ -0,0 +1,592 @@
+
+        chain = ctx->upstream_client_certificate_chain;
+        pkey = ctx->upstream_client_private_key;
+
+        if (sk_X509_num(chain) < 1) {
+            ngx_ssl_error(NGX_LOG_ALERT, c->log, 0,
+                          "invalid client certificate chain provided while "
+                          "handshaking with upstream");
+            goto failed;
+        }
+
+        x509 = sk_X509_value(chain, 0);
+        if (x509 == NULL) {
+            ngx_ssl_error(NGX_LOG_ALERT, c->log, 0, "sk_X509_value() failed");
+            goto failed;
+        }
+
+        if (SSL_use_certificate(sc, x509) == 0) {
+            ngx_ssl_error(NGX_LOG_ALERT, c->log, 0,
+                          "SSL_use_certificate() failed");
+            goto failed;
+        }
+
+        /* read rest of the chain */
+
+        for (i = 1; i < sk_X509_num(chain); i++) {
+            x509 = sk_X509_value(chain, i);
+            if (x509 == NULL) {
+                ngx_ssl_error(NGX_LOG_ALERT, c->log, 0,
+                              "sk_X509_value() failed");
+                goto failed;
+            }
+
+            if (SSL_add1_chain_cert(sc, x509) == 0) {
+                ngx_ssl_error(NGX_LOG_ALERT, c->log, 0,
+                              "SSL_add1_chain_cert() failed");
+                goto failed;
+            }
+        }
+
+        if (SSL_use_PrivateKey(sc, pkey) == 0) {
+            ngx_ssl_error(NGX_LOG_ALERT, c->log, 0,
+                          "SSL_use_PrivateKey() failed");
+            goto failed;
+        }
+    }
+
+    if (ctx->upstream_trusted_store != NULL) {
+        ngx_log_debug0(NGX_LOG_DEBUG_HTTP, c->log, 0,
+                       "overriding upstream SSL trusted store");
+        store = ctx->upstream_trusted_store;
+
+        if (SSL_set1_verify_cert_store(sc, store) == 0) {
+            ngx_ssl_error(NGX_LOG_ALERT, c->log, 0,
+                          "SSL_set1_verify_cert_store() failed");
+            goto failed;
+        }
+    }
+
+    if (ctx->upstream_ssl_verify_depth_set) {
+        ngx_log_debug0(NGX_LOG_DEBUG_HTTP, c->log, 0,
+                       "overriding upstream SSL verify depth");
+        SSL_set_verify_depth(sc, ctx->upstream_ssl_verify_depth);
+    }
+
+
+    return;
+
+failed:
+
+    ERR_clear_error();
+}
+
+
+static X509 *
+ngx_http_lua_kong_x509_copy(X509 *in)
+{
+    return X509_up_ref(in) == 0 ? NULL : in;
+}
+
+
+static void
+ngx_http_lua_kong_cleanup_cert_and_key(ngx_http_lua_kong_ctx_t *ctx)
+{
+    if (ctx->upstream_client_certificate_chain != NULL) {
+        sk_X509_pop_free(ctx->upstream_client_certificate_chain, X509_free);
+        EVP_PKEY_free(ctx->upstream_client_private_key);
+    }
+}
+
+
+static void
+ngx_http_lua_kong_cleanup_trusted_store(ngx_http_lua_kong_ctx_t *ctx)
+{
+    if (ctx->upstream_trusted_store != NULL) {
+        X509_STORE_free(ctx->upstream_trusted_store);
+    }
+}
+
+
+static void
+ngx_http_lua_kong_cleanup(void *data)
```

Comment:
let's add "namespace" for this function, as it's only used inside the ssl part. how about `ngx_http_lua_kong_ssl_cleanup`, so we can avoid collision of symbols.

Reply:
It is used also in grpc, so we must keep this name.